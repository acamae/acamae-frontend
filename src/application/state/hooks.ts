import { useDispatch, useSelector } from 'react-redux';

import type { RootState, AppDispatch } from '@domain/types/redux';

// usar siempre los hooks personalizados (useAppDispatch, useAppSelector) en toda la app,
// y as√≠ evitar tener que tipar manualmente cada uso de useDispatch o useSelector.
export const useAppDispatch = useDispatch.withTypes<AppDispatch>();
export const useAppSelector = useSelector.withTypes<RootState>();
