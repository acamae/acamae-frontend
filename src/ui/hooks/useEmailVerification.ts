import { useCallback, useEffect, useRef, useState } from 'react';

import { VerifyEmailUseCase } from '@application/use-cases/auth/VerifyEmailUseCase';
import { ApiErrorCodes } from '@domain/constants/errorCodes';
import { AuthApiRepository } from '@infrastructure/api/AuthApiRepository';

export type EmailVerificationStatus =
  | 'IDLE'
  | 'MISSING_TOKEN'
  | 'OFFLINE'
  | 'LOADING'
  | 'SUCCESS'
  | 'ERROR';

export interface UseEmailVerificationResult {
  status: EmailVerificationStatus;
  errorCode?: string;
  verify: (token: string | null) => void;
}

/**
 * Hook que encapsula la lógica de verificación de correo.
 * Maneja timeout, offline, reintentos y mapea los diferentes códigos de error.
 */
export const useEmailVerification = (): UseEmailVerificationResult => {
  const [status, setStatus] = useState<EmailVerificationStatus>('IDLE');
  const [errorCode, setErrorCode] = useState<string | undefined>();

  // Evitamos ejecutar varias veces con useRef
  const controllerRef = useRef<AbortController | null>(null);

  const verify = useCallback((token: string | null) => {
    if (!token) {
      setStatus('MISSING_TOKEN');
      return;
    }

    if (!navigator.onLine) {
      setStatus('OFFLINE');
      return;
    }

    // Cancel previous request if any
    controllerRef.current?.abort();
    const abortController = new AbortController();
    controllerRef.current = abortController;

    const authRepository = new AuthApiRepository();
    const useCase = new VerifyEmailUseCase(authRepository);

    setStatus('LOADING');

    const timeoutId = setTimeout(() => {
      abortController.abort();
      setStatus('ERROR');
      setErrorCode(ApiErrorCodes.ETIMEDOUT);
    }, 10000); // 10s timeout

    useCase
      .execute(token)
      .then(response => {
        clearTimeout(timeoutId);
        if (response.success) {
          setStatus('SUCCESS');
          return;
        }
        // Mapear códigos
        switch (response.code) {
          case ApiErrorCodes.AUTH_TOKEN_EXPIRED:
          case ApiErrorCodes.AUTH_TOKEN_INVALID:
          case ApiErrorCodes.AUTH_TOKEN_REVOKED:
          case ApiErrorCodes.AUTH_TOKEN_MALICIOUS:
            setStatus('ERROR');
            setErrorCode(response.code);
            break;
          case ApiErrorCodes.AUTH_TOKEN_ALREADY_USED:
            setStatus('ERROR');
            setErrorCode(response.code);
            break;
          case ApiErrorCodes.AUTH_RATE_LIMIT:
          case ApiErrorCodes.SERVICE_UNAVAILABLE:
            setStatus('ERROR');
            setErrorCode(response.code);
            break;
          case ApiErrorCodes.AUTH_USER_ALREADY_VERIFIED:
            setStatus('ERROR');
            setErrorCode(response.code);
            break;
          case ApiErrorCodes.AUTH_USER_NOT_FOUND:
          case ApiErrorCodes.AUTH_USER_BLOCKED:
            setStatus('ERROR');
            setErrorCode(response.code);
            break;
          default:
            setStatus('ERROR');
            setErrorCode(response.code);
        }
      })
      .catch(error => {
        clearTimeout(timeoutId);
        if (abortController.signal.aborted) return; // ya gestionado

        // Podríamos inspeccionar error.code o error.message
        setStatus('ERROR');
        setErrorCode(error?.code ?? ApiErrorCodes.UNKNOWN_ERROR);
      });
  }, []);

  // Listener para volver a intentar cuando regrese la conexión
  useEffect(() => {
    const onOnline = () => {
      if (status === 'OFFLINE') {
        setStatus('IDLE');
      }
    };
    window.addEventListener('online', onOnline);
    return () => window.removeEventListener('online', onOnline);
  }, [status]);

  return { status, errorCode, verify };
};
