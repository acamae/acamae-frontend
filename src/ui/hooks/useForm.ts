import { useState, ChangeEvent, FormEvent, useEffect } from 'react';
import { useTranslation } from 'react-i18next';

export type UseFormConfig<T> = {
  initialValues: T;
  onSubmit: (values: T) => void;
  validate?: (values: T) => Partial<Record<keyof T, string>>;
};

export type UseFormReturn<T> = {
  values: T;
  errors: Partial<Record<keyof T, string>>;
  touched: Partial<Record<keyof T, boolean>>;
  handleChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  handleSubmit: (config: UseFormConfig<T>) => (e: React.FormEvent) => void;
  isSubmitting?: boolean;
};

export const useForm = <T extends object>({
  initialValues,
  onSubmit,
  validate,
}: UseFormConfig<T>) => {
  const [values, setValues] = useState<T>(initialValues);
  const [errors, setErrors] = useState<Partial<Record<keyof T, string>>>({});
  const [isSubmitting, setIsSubmitting] = useState(false);
  const [touched, setTouched] = useState<Record<keyof T, boolean>>({} as Record<keyof T, boolean>);
  const { i18n } = useTranslation();

  // Run validation only when the language changes and only for touched fields
  useEffect(() => {
    const validationErrors =
      validate && Object.keys(touched).length > 0 ? validate(values) || {} : {};
    setErrors(validationErrors);
  }, [i18n.language, validate, values, touched]);

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    const { name, value } = e.target;
    setValues(
      prev =>
        ({
          ...prev,
          [name]: value,
        }) as T
    );

    // Set the field as touched
    setTouched(
      prev =>
        ({
          ...prev,
          [name]: true,
        }) as Record<keyof T, boolean>
    );

    const validationErrors = validate ? validate({ ...values, [name]: value }) || {} : {};
    setErrors(validationErrors);
  };

  const handleSubmit = async (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    // Set all fields as touched
    const allTouched = Object.keys(values).reduce(
      (acc, key) => ({ ...acc, [key]: true }),
      {} as Record<keyof T, boolean>
    );

    setTouched(allTouched);

    const validationErrors = validate ? validate(values) || {} : {};
    setErrors(validationErrors);

    if (Object.keys(validationErrors).length > 0) {
      return;
    }

    setIsSubmitting(true);

    try {
      await onSubmit(values);
    } catch (error) {
      console.error('Error submitting form:', error);
    } finally {
      setIsSubmitting(false);
    }
  };

  const resetForm = () => {
    setValues(initialValues);
    setErrors({});
    setTouched({} as Record<keyof T, boolean>);
  };

  return {
    values,
    errors,
    isSubmitting,
    touched,
    handleChange,
    handleSubmit,
    resetForm,
  };
};
