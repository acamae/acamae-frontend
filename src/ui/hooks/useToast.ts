import { useToastContext } from '@shared/services/ToastProvider';

export const useToast = () => useToastContext();
export default useToast;
