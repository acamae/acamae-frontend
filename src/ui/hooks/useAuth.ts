import {
  loginAction,
  registerAction,
  logoutAction,
  forgotPasswordAction,
  resetPasswordAction,
} from '@application/state/actions/auth.actions';
import { useAppDispatch, useAppSelector } from '@application/state/hooks';
import {
  ForgotPasswordPayload,
  LoginPayload,
  RegisterPayload,
  ResetPasswordPayload,
} from '@domain/types/apiSchema';
import type { RootState } from '@domain/types/redux';

export const useAuth = () => {
  const dispatch = useAppDispatch();
  const { user, token, isAuthenticated, loading, error } = useAppSelector(
    (state: RootState) => state.auth
  );
  const login = (loginPayload: LoginPayload) => dispatch(loginAction(loginPayload));
  const register = (registerPayload: RegisterPayload) => dispatch(registerAction(registerPayload));
  const forgotPassword = (forgotPasswordPayload: ForgotPasswordPayload) =>
    dispatch(forgotPasswordAction(forgotPasswordPayload));
  const resetPassword = (resetPasswordPayload: ResetPasswordPayload) =>
    dispatch(resetPasswordAction(resetPasswordPayload));
  const logout = () => dispatch(logoutAction());

  return {
    user,
    token,
    isAuthenticated,
    loading,
    error,
    login,
    register,
    logout,
    forgotPassword,
    resetPassword,
  };
};
