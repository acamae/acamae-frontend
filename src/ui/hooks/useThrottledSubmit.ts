import { useState, useCallback, useEffect } from 'react';
import { useTranslation } from 'react-i18next';

import { ThrottleConfig, THROTTLE_CONFIGS } from '@shared/constants/security';
import { securityThrottleService, generateActionId } from '@shared/utils/securityUtils';
import { useToast } from '@ui/hooks/useToast';

export interface UseThrottledSubmitConfig {
  formName: string;
  throttleConfig?: Partial<ThrottleConfig>;
  onSubmit: () => Promise<void>;
  showToastOnThrottle?: boolean;
}

export interface UseThrottledSubmitReturn {
  isThrottled: boolean;
  canSubmit: boolean;
  timeUntilNextSubmission: number;
  remainingAttempts: number;
  handleThrottledSubmit: () => Promise<void>;
  resetThrottle: () => void;
  activateThrottle: () => void;
}

export const useThrottledSubmit = ({
  formName,
  throttleConfig = THROTTLE_CONFIGS.AUTH_FORMS as Partial<ThrottleConfig>,
  onSubmit,
  showToastOnThrottle = true,
}: UseThrottledSubmitConfig): UseThrottledSubmitReturn => {
  const { t } = useTranslation();
  const toast = useToast();
  const [isThrottled, setIsThrottled] = useState(false);
  const [timeUntilNextSubmission, setTimeUntilNextSubmission] = useState(0);
  const [remainingAttempts, setRemainingAttempts] = useState(0);

  const actionId = generateActionId(formName, 'submit');

  // Función para actualizar el estado del throttle
  const updateThrottleState = useCallback(() => {
    const timeRemaining = securityThrottleService.getTimeUntilNextAction(actionId, throttleConfig);
    const attempts = securityThrottleService.getRemainingAttempts(actionId, throttleConfig);
    setTimeUntilNextSubmission(timeRemaining);
    setRemainingAttempts(attempts);
    setIsThrottled(timeRemaining > 0);
  }, [actionId, throttleConfig]);

  // Actualizar el estado inicial al montar el componente
  useEffect(() => {
    updateThrottleState();
  }, [updateThrottleState]);

  // Actualizar el tiempo restante cada segundo
  useEffect(() => {
    const interval = setInterval(() => {
      updateThrottleState();
    }, 1000);

    return () => clearInterval(interval);
  }, [updateThrottleState]);

  const handleThrottledSubmit = useCallback(async (): Promise<void> => {
    // Verificar si la acción está permitida
    if (!securityThrottleService.canPerformAction(actionId, throttleConfig)) {
      // Actualizar el estado inmediatamente después del submit
      updateThrottleState();

      if (showToastOnThrottle) {
        const throttleState = securityThrottleService.getThrottleState(actionId);

        if (throttleState?.isBlocked) {
          // Usuario bloqueado por demasiados intentos
          toast.error(
            t('security.throttle.blocked', {
              minutes: Math.ceil(timeUntilNextSubmission / 60000),
            })
          );
        } else {
          // Usuario debe esperar antes del próximo intento
          toast.warning(
            t('security.throttle.wait', {
              seconds: Math.ceil(timeUntilNextSubmission / 1000),
            })
          );
        }
      }

      return;
    }

    await onSubmit();

    // Actualizar el estado inmediatamente después del submit exitoso
    updateThrottleState();
  }, [
    actionId,
    throttleConfig,
    onSubmit,
    showToastOnThrottle,
    t,
    toast,
    updateThrottleState,
    timeUntilNextSubmission,
  ]);

  const resetThrottle = useCallback(() => {
    securityThrottleService.clearThrottleState(actionId);
    setIsThrottled(false);
    setTimeUntilNextSubmission(0);
  }, [actionId]);

  const activateThrottle = useCallback(() => {
    // Activar el throttling manualmente cuando el servidor devuelve un error 429
    securityThrottleService.recordAction(actionId, throttleConfig);

    // Actualizar el estado inmediatamente
    const timeRemaining = securityThrottleService.getTimeUntilNextAction(actionId, throttleConfig);
    const attempts = securityThrottleService.getRemainingAttempts(actionId, throttleConfig);

    setTimeUntilNextSubmission(timeRemaining);
    setRemainingAttempts(attempts);
    setIsThrottled(true);

    if (showToastOnThrottle) {
      toast.error(t('security.throttle.server_blocked'));
    }
  }, [actionId, throttleConfig, showToastOnThrottle, t, toast]);

  const canSubmit = !isThrottled && timeUntilNextSubmission === 0;

  return {
    isThrottled,
    canSubmit,
    timeUntilNextSubmission,
    remainingAttempts,
    handleThrottledSubmit,
    resetThrottle,
    activateThrottle,
  };
};

/**
 * Hook especializado para formularios de autenticación
 */
export const useAuthThrottledSubmit = (
  formName: string,
  onSubmit: () => Promise<void>
): UseThrottledSubmitReturn => {
  return useThrottledSubmit({
    formName,
    throttleConfig: THROTTLE_CONFIGS.AUTH_FORMS as Partial<ThrottleConfig>,
    onSubmit,
    showToastOnThrottle: true,
  });
};

/**
 * Hook especializado para formularios regulares
 */
export const useRegularThrottledSubmit = (
  formName: string,
  onSubmit: () => Promise<void>
): UseThrottledSubmitReturn => {
  return useThrottledSubmit({
    formName,
    throttleConfig: THROTTLE_CONFIGS.REGULAR_FORMS as Partial<ThrottleConfig>,
    onSubmit,
    showToastOnThrottle: true,
  });
};

/**
 * Hook especializado para acciones críticas
 */
export const useCriticalThrottledSubmit = (
  formName: string,
  onSubmit: () => Promise<void>
): UseThrottledSubmitReturn => {
  return useThrottledSubmit({
    formName,
    throttleConfig: THROTTLE_CONFIGS.CRITICAL_ACTIONS as Partial<ThrottleConfig>,
    onSubmit,
    showToastOnThrottle: true,
  });
};
