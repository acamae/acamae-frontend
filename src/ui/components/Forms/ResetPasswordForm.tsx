import React, { useState, useCallback } from 'react';
import { Form, Button, InputGroup, Alert } from 'react-bootstrap';
import { useTranslation } from 'react-i18next';

import { ApiResponse } from '@/domain/types/apiSchema';
import { validatePassword } from '@domain/services/validationService';
import { useAuth } from '@ui/hooks/useAuth';
import { useForm } from '@ui/hooks/useForm';
import { useToast } from '@ui/hooks/useToast';

import PasswordStrengthMeter from './PasswordStrengthMeter';

interface ResetPasswordFormProps {
  token: string;
}

const ResetPasswordForm: React.FC<ResetPasswordFormProps> = ({ token = '' }) => {
  const { t } = useTranslation();
  const { resetPassword, loading, error } = useAuth();
  const toast = useToast();
  const [showPassword, setShowPassword] = useState(false);
  const [message, setMessage] = useState<string | null>(null);
  const [isSuccess, setIsSuccess] = useState<boolean | null>(null);

  const { values, errors, touched, handleChange, handleSubmit, isSubmitting } = useForm<{
    password: string;
  }>({
    initialValues: { password: '' },

    validate: useCallback(
      (values: { password: string }) => {
        const errors: Partial<{ password: string }> = {};
        if (!values.password) {
          errors.password = t('errors.password.required');
        } else if (!validatePassword(values.password)) {
          errors.password = t('errors.password.invalid');
        }
        return errors;
      },
      [t]
    ),

    onSubmit: async ({ password }) => {
      if (!token) {
        setMessage(t('reset.invalid_token'));
        setIsSuccess(false);
        return;
      }
      const action = await resetPassword({ token, newPassword: password });
      const result = action.payload as ApiResponse<void>;
      setIsSuccess(!error);
      setMessage(error ? t('reset.failed') : t('reset.success'));

      if (!error) {
        toast.success(result?.message || t('reset.success'));
      } else {
        toast.error(result?.message || t('reset.failed'));
      }
    },
  });

  if (message) {
    return (
      <Alert
        variant={isSuccess ? 'success' : 'danger'}
        role={isSuccess ? 'status' : 'alert'}
        aria-live="polite"
        data-testid={isSuccess ? 'alert-success' : 'alert-error'}>
        {message}
      </Alert>
    );
  }

  return (
    <>
      {token ? (
        <Form onSubmit={handleSubmit} noValidate>
          <Form.Group className="mb-3" controlId="password">
            <Form.Label data-testid="label-reset-password">{t('reset.password')}</Form.Label>
            <InputGroup hasValidation>
              <Form.Control
                type={showPassword ? 'text' : 'password'}
                name="password"
                value={values.password}
                onChange={handleChange}
                isInvalid={touched.password && !!errors.password}
                required
                autoComplete="new-password"
                aria-describedby="passwordHelp"
                data-testid="input-password"
              />
              <Button
                variant="outline-secondary"
                onClick={() => setShowPassword(prev => !prev)}
                aria-label={t('register.toggle_password')}
                data-testid="btn-toggle-password"
                tabIndex={-1}>
                {showPassword ? 'üôà' : 'üëÅÔ∏è'}
              </Button>
              <Form.Control.Feedback type="invalid">{errors.password}</Form.Control.Feedback>
            </InputGroup>
            <Form.Text id="passwordHelp" className="text-muted">
              {t('register.password_help')}
            </Form.Text>
            <PasswordStrengthMeter password={values.password} t={t} />
          </Form.Group>
          <div className="d-grid">
            <Button
              variant="primary"
              type="submit"
              disabled={isSubmitting || loading}
              data-testid="btn-reset-password">
              {isSubmitting || loading ? t('reset.saving') : t('reset.submit')}
            </Button>
          </div>
        </Form>
      ) : (
        <Alert variant="danger" role="status" aria-live="polite" data-testid="alert-error">
          {t('reset.invalid_token')}
        </Alert>
      )}
    </>
  );
};

export default ResetPasswordForm;
