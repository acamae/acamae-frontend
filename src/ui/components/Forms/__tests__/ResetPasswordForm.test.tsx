jest.mock('@ui/hooks/useAuth');
jest.mock('@ui/hooks/useToast');
jest.mock('react-i18next');

import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { useTranslation } from 'react-i18next';

import ResetPasswordForm from '@ui/components/Forms/ResetPasswordForm';
import { useAuth } from '@ui/hooks/useAuth';
import { useToast } from '@ui/hooks/useToast';

function setupUseAuth({ loading = false, resetPassword = jest.fn() } = {}) {
  (useAuth as jest.Mock).mockReturnValue({
    loading,
    resetPassword,
  });
  return { resetPassword };
}

function setupUseToast({ success = jest.fn(), error = jest.fn() } = {}) {
  (useToast as jest.Mock).mockReturnValue({ success, error });
  return { success, error };
}

const changeLanguageMock = jest.fn();
function setupUseTranslation({
  lang = 'es-ES',
  t = (str: string) => str,
  changeLanguage = changeLanguageMock,
} = {}) {
  (useTranslation as jest.Mock).mockReturnValue({
    t,
    i18n: {
      language: lang,
      changeLanguage,
    },
  });
}

function renderResetPasswordForm(props = { token: 'mock-token' }) {
  return render(<ResetPasswordForm {...props} />);
}

describe('ResetPasswordForm', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    setupUseTranslation();
    setupUseToast();
    setupUseAuth();
  });

  it('renderiza correctamente los campos', () => {
    renderResetPasswordForm();
    expect(screen.getByTestId('input-password')).toBeInTheDocument();
    expect(screen.getByTestId('btn-reset-password')).toBeInTheDocument();
  });

  it('muestra error de validación si el password es inválido', () => {
    renderResetPasswordForm();
    fireEvent.change(screen.getByTestId('input-password'), {
      target: { value: '' },
    });
    fireEvent.click(screen.getByTestId('btn-reset-password'));
    expect(screen.getByText('errors.password.required')).toBeInTheDocument();
  });

  it('muestra mensaje de éxito tras submit correcto', async () => {
    const resetPassword = jest.fn().mockResolvedValue({
      payload: { success: true, message: 'reset.password_updated' },
    });
    setupUseAuth({ resetPassword });
    renderResetPasswordForm();
    fireEvent.change(screen.getByTestId('input-password'), {
      target: { value: 'Password123' },
    });
    fireEvent.click(screen.getByTestId('btn-reset-password'));
    await waitFor(() => {
      expect(screen.getByTestId('alert-success')).toBeInTheDocument();
      expect(screen.getByTestId('alert-success')).toHaveTextContent('reset.success');
    });
  });

  it('muestra mensaje de error si el token es inválido', async () => {
    renderResetPasswordForm({ token: '' });
    expect(screen.getByTestId('alert-error')).toBeInTheDocument();
    expect(screen.getByTestId('alert-error')).toHaveTextContent('reset.invalid_token');
  });

  it('deshabilita el botón de submit mientras loading es true', () => {
    setupUseAuth({ loading: true });
    renderResetPasswordForm();
    expect(screen.getByTestId('btn-reset-password')).toBeDisabled();
  });

  it('deshabilita el botón de submit mientras isSubmitting es true', () => {
    setupUseAuth({ loading: true });
    renderResetPasswordForm();
    expect(screen.getByTestId('btn-reset-password')).toBeDisabled();
  });

  it('snapshot render', () => {
    const { asFragment } = renderResetPasswordForm();
    expect(asFragment()).toMatchSnapshot();
  });
});
