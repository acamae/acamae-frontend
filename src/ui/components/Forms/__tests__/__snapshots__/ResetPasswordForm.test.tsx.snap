// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetPasswordForm snapshot render 1`] = `
<DocumentFragment>
  <form
    class=""
    novalidate=""
  >
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        data-testid="label-reset-password"
        for="password"
      >
        reset.password
      </label>
      <div
        class="input-group has-validation"
      >
        <input
          aria-describedby="passwordHelp"
          autocomplete="new-password"
          class="form-control"
          data-testid="input-password"
          id="password"
          name="password"
          required=""
          type="password"
          value=""
        />
        <button
          aria-label="register.toggle_password"
          class="btn btn-outline-secondary"
          data-testid="btn-toggle-password"
          tabindex="-1"
          type="button"
        >
          👁️
        </button>
        <div
          class="invalid-feedback"
        />
      </div>
      <small
        class="text-muted form-text"
        id="passwordHelp"
      >
        register.password_help
      </small>
      <div
        class="form-text"
      >
        <span
          data-testid="password-strength-label"
        >
          register.strength.weak
        </span>
        <div
          aria-hidden="true"
          class="progress mt-1"
        >
          <div
            class="progress-bar bg-danger progressBar progressBar-1"
            data-testid="password-strength-bar"
          />
        </div>
      </div>
    </div>
    <div
      class="d-grid"
    >
      <button
        class="btn btn-primary"
        data-testid="btn-reset-password"
        type="submit"
      >
        reset.submit
      </button>
    </div>
  </form>
</DocumentFragment>
`;
